studio/studio.py

```diff
@@ -237,6 +245,171 @@ async def export(q: Q):
+    ]
+    await q.page.save()
+
+
+async def block_dialog(q: Q):
+    q.page['meta'].dialog.closable = False
+    q.page['meta'].dialog.blocking = True
+    await q.page.save()
+
+
+async def on_pip_finish(q: Q):
+    q.page['meta'].dialog.blocking = False
+    q.page['meta'].dialog.closable = True
+    q.page['meta'].dialog.items = get_package_dialog_items()
+    await q.page.save()
+
+
+async def pip(q: Q, command: str, flag: str or None, package_name: str, package_version: str, progress_message: str,
+              success_message: str, error_message: str, on_success: Callable = None, on_error: Callable = None):
+    try:
+        version = f'=={package_version}' if package_version else ''
+        args = [sys.executable, '-m', 'pip', command, flag, f'{package_name}{version}']
+        p = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output = ''
+        await block_dialog(q)
+        await show_progress(q, progress_message)
+        while True:
+            if p.poll() is not None:
+                # Process finished.
+                if p.returncode == 0:
+                    if on_success:
+                        await on_success(package_name)
+                    await show_finish_message(q, 'success', success_message, output)
+                else:
+                    if on_error:
+                        await on_error()
+                    output += f'\n{p.stderr.read().decode("utf-8")}'
+                    await show_finish_message(q, 'error', error_message, output)
+                break
+            else:
+                output += f'\n{p.stdout.readline().decode("utf-8")}'
+                await update_progress(q, output)
+    finally:
+        if p.poll() is None:
+            # Task canceled.
+            p.stdout.close()
+            p.stderr.close()
+            p.kill()
+            await on_pip_finish(q)
+
+
+def update_requirements(packages: dict, remove=False):
+    with open('project/requirements.txt', 'a+') as file:
+        file.seek(0)
+        lines = file.readlines()
+
+    current_packages = {}
+    for line in lines:
+        package_name, package_version = line.strip().split('==')
+        current_packages[package_name] = package_version
+
+    if not remove:
+        current_packages.update(packages)
+
+    with open('project/requirements.txt', 'w') as file:
+        for package_name, package_version in current_packages.items():
+            if not remove or (remove and package_name not in packages):
+                file.write(f'{package_name}=={package_version}\n')
+
+
+async def on_install_success(package_name: str):
+    update_requirements({f'{package_name}': version(package_name)})
+
+
+async def on_uninstall_success(package_name: str):
+    update_requirements({f'{package_name}': ''}, True)
+
+
+async def on_requirements_install_success(package_name: str):
+    packages = {}
+    with open('project/requirements_tmp.txt', 'r') as file_tmp:
```

Comment:
Why do we need tmp requirements?

Reply:
The original implementation was for uploaded `requirements.txt` to update existing one and this was preventing of the file overwrite. 

Removing and adding `Packages from requirements.txt will replace currently installed packages.` into file upload tooltip.